Minimum required patches for building ncurses libraries on Visual Studio

madebr opened the pull request:
These patches are the minimum required patches to let me build the ncurses libraries on Windows using Visual Studio.

I created these while trying to package ncurses 6.2 for conan.
My WIP scripts can be found at https://github.com/madebr/conan-center-index/tree/ncurses/recipes/ncurses/all

Building for Visual Studio was causing some problems because it is missing some features.

This patchset adds workarounds for building the ncurses libraries.

What these patches do not fix is biuilding shared ncurses shared libraries (dlls) on Windows.
`NCURSES_IMPEXP` is defined globally and is the same for every ncurses sublibrary.
Whereas each sublibrary must have its own definition.
What I mean is that when building libmenu, a preprocessor variable `LIBMENU_IMPEXP` must be `__declspec(dllexport)` but `NCURSES_IMPEXP` must be `__declspec(dllimport)`.

ThomasDickey commented:
The [`NOMACROS`](https://github.com/mirror/ncurses/pull/1/commits/a22f9ec63b169aeec4b0d45df696a1ae8b358470) change will cause the library to not work...

madebr commented:
@ThomasDickey 
I've got to admit that I did not give much attention to fixing the errors in that patch.
I'll rebase my paches on current master and see how things changed.

ThomasDickey commented:
I haven't tested with msvc yet, but current code seems to work with mingw.  Compiler warning made me notice an error in one `memcpy` (fixed), but the rest of my changes for `win_driver.c` were to reduce compiler warnings.  So I don't think there'll be much surprise - just something to review, etc.

madebr commented:
I've updated the patches.
Windows CI output at https://ci.appveyor.com/project/madebr/ncurses/builds/31521971

The `NOMACROS` patch is still there because of an error.
Please read the commit message to see what error is thrown.
https://github.com/mirror/ncurses/pull/1/commits/6e407c1408198874b57095fd46220b99024ae20d
Maybe you can immediately see what's wrong.

The same goes for the other error:
https://github.com/mirror/ncurses/pull/1/commits/a31f13248c86db9cb27aba007dbd34628caee4bf

ThomasDickey commented:
yes, I'm aware of this, but the lib_gen.c change will not produce a working library (just something that compiles).  I've been setting up a development environment to work on this, but am not done.

madebr commented:
> 
> 
> yes, I'm aware of this, but the lib_gen.c change will not produce a working library (just something that compiles). I've been setting up a development environment to work on this, but am not done.

I've taken a look at it again and I fixed it in 'Fix lib_gen.c'. I think.

ThomasDickey commented:
Defining `NCURSES_NOMACROS` will allow it to compile, but the library will not work, because **calling** those functions will result in an infinite recursion on the stack.

madebr commented:
Currently required:
- for shared libraries, use a define for each library
In CMake terminology, this is using [DEFINE_SYMBOL](https://cmake.org/cmake/help/latest/prop_tgt/DEFINE_SYMBOL.html) + dllimport/dllexport.
This is done to e.g. avoid the form library to export a function, defined in ncurses.
- fix `lib_gen.c` by modifying the regex in `ncurses/base/MKlib_gen.sh`

appveyour ci: https://ci.appveyor.com/project/madebr/ncurses/builds/33052220

TODO (What is not fixed/untouched)
- the progs cannot be built because getopt is not available on MSVC, so configure fails.
Using https://github.com/skandhurkat/Getopt-for-Visual-Studio (or similar) does not work either because configure.in tests for the existence of `getopt` without a header, which it does not on MSVC (it requires `#include <getopt.h>`)
- the progs do not link on MSVC, because the arguments are incompatible between gcc/msvc.

ThomasDickey commented:
Referring to the [packaging_patches](https://github.com/madebr/ncurses/tree/packaging_patches) branch, some of those have been addressed:

- [Fix lib_gen.c ](https://github.com/madebr/ncurses/commit/effaed661e964f5fe3cb777e6bdb5f5450a77cb8)
- [Add include file to add export annotation ](https://github.com/madebr/ncurses/commit/e8ab3ab7e855920ae2c617fd7306909ef621cbda)

These are the ones that I'm considering:

- [Pass BUILDING_XXX compile definition while building library XXX ](https://github.com/madebr/ncurses/commit/b8400a1b3ba8a00cd49783799d1c3b9576b02878)
- [Let each sublibrary do its own import/exporting ](https://github.com/madebr/ncurses/commit/0f571649d893e8d7478f90506ba9ff5dfc796cd9)

The problem that I see in these is that they change longstanding public-use header files, making them dependent up a different build configuration.  I'm considering how to solve the problem without breaking compatibility.

This patch is solving a problem which probably should be done outside ncurses:

- [Enable building progs using shared MSVC ](https://github.com/madebr/ncurses/commit/b24a45d3c622f14f7720e2b01327ba3a33a1a711)

These don't look suitable for inclusion in ncurses:

- [Added .gitignore ](https://github.com/madebr/ncurses/commit/558ec611e1beb08050a7648c1999eddf84e05b68)
- [Add appveyor CI ](https://github.com/madebr/ncurses/commit/cf9de8bdf21c4771d939bc6d92cdec20ffbdf546)

madebr commented:
> Referring to the [packaging_patches](https://github.com/madebr/ncurses/tree/packaging_patches) branch, some of those have been addressed:
> 
>     * [Fix lib_gen.c ](https://github.com/madebr/ncurses/commit/effaed661e964f5fe3cb777e6bdb5f5450a77cb8) (though I recall that one of the later patches relies upon the syntax change)

See https://github.com/mirror/ncurses/commit/c976a90788f3e50afc773670ff74c5270ecb48df#diff-f425a63d0f18f4bbd666c1086bac3b6d
This script requires 2 changes in the 6.2 release.

> The problem that I see in these is that they change longstanding public-use header files, making them dependent up a different build configuration. I'm considering how to solve the problem without breaking compatibility.

Using MSVC, these changes can be avoided by using module-definition (.def) files.
See https://docs.microsoft.com/en-us/cpp/build/exporting-from-a-dll-using-def-files?view=vs-2019
This would require listing all exported symbols for ncurses, and for all its sublibraries, in a file.
But this is not really advised, and bug prone because symbols can be missed easily.

See https://www.gnu.org/software/gnulib/manual/html_node/Exported-Symbols-of-Shared-Libraries.html#Exported-Symbols-of-Shared-Libraries,
where the same process as I have used in these patches is described.

> This patch is solving a problem which probably should be done outside ncurses:
> 
>     * [Enable building progs using shared MSVC ](https://github.com/madebr/ncurses/commit/b24a45d3c622f14f7720e2b01327ba3a33a1a711)

Yes, because this patch creates a script that will link object files to an executable.
No, because Makefile also passes `LDFLAGS` (such as `-link -dll` that are meant only for linking dll's).
I don't see how I can externally override `LINK_PROGS` if this script should be kept outside of ncurses.

> These don't look suitable for inclusion in ncurses:
> 
>     * [Added .gitignore ](https://github.com/madebr/ncurses/commit/558ec611e1beb08050a7648c1999eddf84e05b68)
> 
>     * [Add appveyor CI ](https://github.com/madebr/ncurses/commit/cf9de8bdf21c4771d939bc6d92cdec20ffbdf546)

I totally agree.
I added these as a proof that these patches make ncurses compile on MSVC.
(please continue to ignore these)

ThomasDickey commented:
hmm - at the moment I'm test-building the mingw stuff, to review, etc. (I didn't find time to test with msvc, too many distractions recently).

ThomasDickey commented:
Today I tried the full set of changes, noticed redefinition warnings (in the import/export stuff), backed out those changes for now, but added the `aclocal.m4` change and amended the `lib_gen.c` change (the newline substitution isn't portable, but the parentheses don't appear to be needed now).

madebr commented:
Cool!
On what os/compiler + build type are these warnings appearing?
I only tested the common combinations, available in the conanfile: normal and shared.

While you're looking at exporting symbols on mingw,
I think these patches will break building a dll using libtool because `NCURSES_STATIC` is defined for it. I don't know how to make CFLAGS dependent on the *shared'ness* of the library using libtool.

ThomasDickey commented:
I was using gcc-10 on Debian/testing.  Regarding libtool, I have a different to-do (fixing build problems with the gnat/Ada95), but I mostly build without libtool.  However I have a scripted build for that configuration which I work on.   The `NCURSES_STATIC` does look like a potential problem, but at the moment I was seeing how to reduce this backlog a little (there's also the two bug reports that I'm working on...).

madebr commented:
>  (there's also the two bug reports that I'm working on...).

Please don't feel obligated to work under a deadline to fix this problem..
I'm just posting my patches and findings here so they aren't lost and hopefully end up in ncurses.

Through debugging, I found out what caused the `tic.exe` problem, described in https://github.com/mirror/ncurses/pull/1#issuecomment-632744816.
The `access` function by MSVC only accepts access_mode flags where `R_OK` and/or `W_OK` is set.
The assertion `(access_mode & (~6)) == 0` fails. (`R_OK | W_OK == 6`)
So the flag `X_OK` is the cause of this issue.

The [msdn documentation](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/access-waccess?view=vs-2019) for access only describes R_OK and W_OK.
It also describes the function does not work on directories.

Looking at the [definition of access](https://pubs.opengroup.org/onlinepubs/9699919799/functions/access.html), it looks like its usage is discouraged. So perhaps its use should be reduced (or even removed).

Patching out the use of `X_OK` in a bad patch, the progs build and install fine on Windows.
https://ci.appveyor.com/project/madebr/ncurses/builds/33162174

ThomasDickey commented:
For `access`, I'd wrap that in a macro (as for instance done for binary-mode fopen).  I'm not working this item on a *deadline*, but keep in mind that things that aren't active tend to go onto a continually-growing to-do list :-)

madebr commented:
In https://github.com/mirror/ncurses/pull/1#issuecomment-633157515, you were talking about errors.
I just tried building ncurses with `../configure --with-shared --without-normal --without-debug --enable-fvisibility` and then indeed linking the progs fail because they are using private functions from the ncurses library.
Removing `--enable-fvisibility` makes the library build and link fine (using the configure flags above).

ThomasDickey commented:
In today's patch, I'm adding most of this, except:
- the content of "ncurses_exports.h" is in the pre-existing "ncurses_dll.h"
- the newline-change for MKlib_gen.sh doesn't port
- some import/export declarations were overlooked (added those for building with MinGW).

So you might want to merge and see what remaining changes (such as "access") should be proposed.

madebr commented:
Hello! Great patch weekend!
This indeed allows one to build a ncurses dll using MSVC (with the access patch)!

I've updated the `ACCESS` patch to make it acceptable for upstream.

Unfortunately, this is not the end of the line...
I've added some patches that will fix some problems with building ncurses with wide character support + building the tests.
I was not able (yet) to:
- build a shared ncurses library with wide character support;
- build the tests, linked to a shared ncurses library

Output of CI can be found at https://ci.appveyor.com/project/madebr/ncurses/builds/33921294

---

- Until now, I manually added `user32.lib` to `LDFLAGS` prior to executing`configure`.
Maybe the configure script should automatically add this system library when using Visual Studio?

Adding this manually, fixes the following link error.
```
 Creating library ../lib/ncurses.dll.lib and object ../lib/ncurses.dll.exp
win_driver.obj : error LNK2019: unresolved external symbol __imp_MessageBeep referenced in function wcon_dobeepflash
../lib/ncurses6.dll : fatal error LNK1120: 1 unresolved externals  
```

- Visual Studio does not have a `<search.h>` implementation. This is used when building ncurses with wide character support.
I can fix this by using an [alternative (header-only) implementation](https://github.com/kulp/naive-tsearch). (This is also how I fix a missing getopt)
The test in configure fails nonetheless because it assumes linking to tsearch (and not including `<search.h>`).
The result of this test is also not used in `ncurses/base/new_pair.c`.

(the `with_widec` tests do not work currently, because I've not sent this library to conan for inclusion yet)

- Tests do not build on MSVC: due to missing `struct timeval`.
I fixed this by moving the include of `<winsock.h>` from `ncurses/win32con/gettimeofday.c` to `include/nc_mingw.h`.

This fixes this kind of error:
```
C:/users/maarten/.conan/data/ncurses/6.2-20200704/_/_/build/911e47335817cd75de61d9b47c4d68a137801d08/test/color_content.c(62): error C2079: 'initial_time' uses undefined struct 'timeval'
C:/users/maarten/.conan/data/ncurses/6.2-20200704/_/_/build/911e47335817cd75de61d9b47c4d68a137801d08/test/color_content.c(63): error C2079: 'finish_time' uses undefined struct 'timeval'
C:/users/maarten/.conan/data/ncurses/6.2-20200704/_/_/build/911e47335817cd75de61d9b47c4d68a137801d08/test/color_content.c(178): warning C4133: 'function': incompatible types - from 'int *' to 'timeval *'
C:/users/maarten/.conan/data/ncurses/6.2-20200704/_/_/build/911e47335817cd75de61d9b47c4d68a137801d08/test/color_content.c(219): error C2037: left of 'tv_sec' specifies undefined struct/union 'timeval'
C:/users/maarten/.conan/data/ncurses/6.2-20200704/_/_/build/911e47335817cd75de61d9b47c4d68a137801d08/test/color_content.c(220): error C2037: left of 'tv_usec' specifies undefined struct/union 'timeval'
C:/users/maarten/.conan/data/ncurses/6.2-20200704/_/_/build/911e47335817cd75de61d9b47c4d68a137801d08/test/color_content.c(322): warning C4133: 'function': incompatible types - from 'int *' to 'timeval *'
C:/users/maarten/.conan/data/ncurses/6.2-20200704/_/_/build/911e47335817cd75de61d9b47c4d68a137801d08/test/color_content.c(324): warning C4133: 'function': incompatible types - from 'int *' to 'timeval *'
C:/users/maarten/.conan/data/ncurses/6.2-20200704/_/_/build/911e47335817cd75de61d9b47c4d68a137801d08/test/color_content.c(325): warning C4133: 'function': incompatible types - from 'int *' to 'timeval *'
```

- I fixed linking the tests by copy/pasting `@LINK_PROGS@ $(LIBTOOL_LINK)` from `progs/Makefile.in` to `test/Makefile.in`.

- Building a wide character library fails because of missing `wmemchr`/`wmemcpy` when linking.
This can be fixed by including `<wchar.h>`.
I don't know whether this is okay, because there is a comment about a broken implementation on mingw in `` ncurses/widechar/widechars.c`. Is this still the case? Shouldn't `configure.in` test for working/broken implementations and conditionally use the override?

This fixes these errors while linking:
ncursesw.lib(lib_cchar.obj) : error LNK2019: unresolved external symbol wmemchr referenced in function getcchar
ncursesw.lib(lib_cchar.obj) : error LNK2019: unresolved external symbol wmemcpy referenced in function getcchar


- `test/picsmap.c` uses `popen`/`pclose`, which should be `_popen` and `_pclose` using MSVC.
I fixed this using a `#ifdef`.
- `test/picsmap.c` uses `strcasecmp`.
Use `stricmp` instead.
There was already a comment about portability.
See https://www.ibm.com/support/knowledgecenter/en/ssw_ibm_i_71/rtref/stricmp.htm
`The stricmp function is available for C++ programs. It is available for C only when the program defines the __cplusplus__strings__ macro.`


- `c++/demo.cc` was compiled in the wrong directory which caused the linking to fail because it could not find the `demo.o` object.

- when using static libraries, c++/demo.exe did not link because it could not find the entry point. This was fixed by passing `/subsystem:console` to the linker when creating progs.

I'm not sure the principle of using the `main` function in `cursesmain.cc` as entry point when building a shared `ncurses++.dll` library will work. I think the entry point must be found before loading all libraries the program links to.

---
I was not able to fix these problems. I would love to receive some feedback on how to satisfactory fix these.

- When building a shared ncurses library, linking the tests fails because of a missing `_nc_gettimeofday` symbol.

- When building ncurses using wide characters, `wcwidth` is used in the menu library, but on windows this provided by `_nc_wcwidth`. Which is internal to cursesw.dll and not exported.
This causes the following error while building menuw.dll
```
Creating library ../lib/menuw.dll.lib and object ../lib/menuw.dll.exp
m_global.obj : error LNK2019: unresolved external symbol _nc_wcwidth referenced in function ComputeMaximum_NameDesc_Lengths
```

madebr commented on test/picsmap.c:
```diff
@@ -962,7 +970,7 @@ lookup_rgb(const char *name)
     if (rgb_table != 0) {
 	int n;
 	for (n = 0; rgb_table[n].name != 0; ++n) {
-	    if (!strcasecmp(name, rgb_table[n].name)) {
+	    if (!stricmp(name, rgb_table[n].name)) {
```
This change slept though.
`strcasecmp` is unknown on MSVC.

ThomasDickey commented:
That's a lot of questions -- I'll pick through the list and try to reduce it somewhat.  Offhand:

* `_nc_wcwidth` could be exported (to the menu library, etc., since that uses `curses.priv.h`)
* MinGW and `wchar.h`?  I have a hunch that's not changed (MinGW has not changed much recently).
* popen/pclose could be done with a macro in `test.priv.h` (or even a configure-check, though any POSIX system has those).
* for user32.lib, I'll have to check on what's a suitable tradeoff (lots of special-cases for a compiler tends to become a problem in maintaining the configuration).

ThomasDickey commented:
In today's patch, I (think I) addressed these issues.  However, I see that my native-MinGW builds fail in linking the test-programs (from last week's update), so I'll be making some fixes in that area.

madebr commented:
Indeed, a lot of problems are fixed in this last patch release.
I've added some patches here that allows to build/run all using MSVC, including wide characters.
- the tests also need to pass `NCURSES_STATIC` when building a static ncurses library
- the object file of the c++ test needs to be built in the correct folder. (Looking at my patches now, only the last patch is the correct one)
- in the tests folder, there was a place left where the build system assumes all WIN32 implementations have a flawed wchar implementation.
- The biggest patch is in c++:
  - A main function can not be imported from a dll function. That's why I renamed the main function when on `_WIN32`. This renaming is actually only required when building a dll.
  - static variables of a dll exported class are not exported. That's why uses of these private objects in header files would cause missing symbols because of inlining. Therefore I moved *all* (maybe) uses of these private variables to the `.cc` files. I could not add `NCURSES_CXX_IMPEXP` to these variables because MSVC complained with [error C2487](https://docs.microsoft.com/en-us/cpp/error-messages/compiler-errors-1/compiler-error-c2487)
    The biggest/weirdest change because of this reason is for `titleWindow`,
This static class variable is declared protected. To avoid breakage, I added an extra function that returned this variable. This function returns a reference to a pointer. This allows one to change the pointer.

ThomasDickey commented:
I see (am reviewing the changes now).

Adding `NCURSES_STATIC` to libtool is a problem, since libtool is used to generate shared libraries.  The normal, debug and profile libraries are static.

I found that not adding NCURSES_STATIC to the test-makefile in my native MinGW build was the problem I had from last week, so I filled out `NCURSES_STATIC` in the places where it seems to address your changes.

madebr commented:
> Adding `NCURSES_STATIC` to libtool is a problem, since libtool is used to generate shared libraries. The normal, debug and profile libraries are static.

I understand. Sorry for that. It would be nice if libtool had some way to divert options to shared/static builds.

> I found that not adding NCURSES_STATIC to the test-makefile in my native MinGW build was the problem I had from last week, so I filled out `NCURSES_STATIC` in the places where it seems to address your changes.

You probably saw a lot of `__imp_XXXX` missing symbol errors. That's a callsign for a missing define to hide the `dllimport`.

> If conan doesn't have `tsearch`, then the ncurses library would have had an error if you had tried to use extended-colors. I don't see that in your build-logs (but there's a nuisance warning about `KEY_EVENT` which I may improve).

Indeed, I did not pass `--enable-ext-colors` to the configure script. I guess it is enabled by default?
Doing `--enable-widec --disable-ext-colors` succeeds, but `--enable-widec --enable-ext-colors` fails because of the missing `tsearch`. I've modified my test scripts.

Anyway, only a small change to `aclocal.m4` is required to make sure that `LINK_TESTS` is set correctly for msvc. Without it, the tests fail to link.
Logs are available at https://ci.appveyor.com/project/madebr/ncurses/builds/34182134

Apart from the test fix, I don't have any further problems with building ncurses on MSVC.
Getting rid of the `KEY_EVENT` warning would be very nice.

Do you know of any configuration options that I may not have tested and could cause problems on MSVC?

ThomasDickey commented:
I don't see anything obvious overlooked.  In yesterday's update, I only saw one place (removing CFLAGS from the linker in the test-directory) which I didn't see as necessary.

madebr commented:
I think the `user32.lib` library has not been addressed yet.

Ideally, it should be added to [BUILD_LDFLAGS](https://github.com/mirror/ncurses/blob/cf6a62567b2365c8678b7d561845bdbd1739e5da/ncurses/Makefile.in#L112) when built as a shared library, and added to `LDFLAGS` when built as a static library.

But adding it only to `LDFLAGS` is easier, I guess.

ThomasDickey commented:
I'm puzzled regarding the `BUILD_LDFLAGS` vs `LDFLAGS` comment.  `BUILD_LDFLAGS` is used when linking the build-time utilities (`make_keys`, `make_hash`), while `LDFLAGS` is used for linking the programs such as `tic`.  Of course `BUILD_LDFLAGS` defaults to `$LDFLAGS`, mainly because autoconf isn't really designed for configuring two things at the same time.

madebr commented:
I made that comment because, when creating shared libraries, only `ncurses.dll` should link to user32.dll.
When creating a static library, all executables linking to `ncurses.lib`, should also link to user32.lib.

ThomasDickey commented:
I'm puzzled regarding the `BUILD_LDFLAGS` vs `LDFLAGS` comment.  `BUILD_LDFLAGS` is used when linking the build-time utilities (`make_keys`, `make_hash`), while `LDFLAGS` is used for linking the programs such as `tic`.  Of course `BUILD_LDFLAGS` defaults to `$LDFLAGS`, mainly because autoconf isn't really designed for configuring two things at the same time.

madebr commented:
> ... it's not clear to me if workarounds such as tsearch are part of your current builds).

The conanfile contains the complete build script.
Apart from the tseach workaround, the only thing that attracts my eyes are the following lines:
```python
        if self.settings.os == "Windows":
            self._autotools.defines.append("_WIN32")
            if self.settings.arch == "x86_64":
                self._autotools.defines.append("_WIN64")
```
I'm not sure whether these are really needed.
I have copied them from an earlier conan build script, and didn't check further.
Once I've started my Windows box, I'll let you know.

madebr commented:
appveyor ci of this pr should be available at https://ci.appveyor.com/project/madebr/ncurses/builds/34625902

- While adding the Visual Studio patches to conan at https://github.com/conan-io/conan-center-index/pull/2351,
the CI error'ed when using Visual Studio 2019 and MT (=static non-debug) runtime .
I had to modify the `mk_prog.sh` script to strip out `-MT`,`-MD`,`-MTd` and `-MDd`.

- I could remove `_WIN32` and `_WIN64`, without change of behavior. I also removed it from the conan recipe.
So it will be tested on all MSVC relevant configuration from 2015 until 2019.

madebr commented on aclocal.m4:
```diff
@@ -6914,6 +6914,9 @@ CF_EOF
 				-link)
 					# ignore -link argument
 					;;
+				-MT | MD | -MTd | -MDd)
```
```suggestion
				-MT | -MD | -MTd | -MDd)
```

madebr commented on configure:
```diff
@@ -6037,6 +6037,9 @@ CF_EOF
 				-link)
 					# ignore -link argument
 					;;
+				-MT | MD | -MTd | -MDd)
```
```suggestion
				-MT | -MD | -MTd | -MDd)
```

madebr commented:
I just wanted to point out that all jobs on appveyor are now succeeding as Conan now has a tsearch implementation available.

ThomasDickey commented:
sounds good.

The case statement would be simpler as
`-M[TD] | -M[TD]d)`
